### PR TITLE
KDC setup for Linux

### PR DESCRIPTION
**Pull request by rahulkotecha**
cc: @stephentoub, @bartonjs, @Priya91 

- Added KDC setup script and config files (tested on Ubuntu, Debian, CentOS and Red Hat)
- Changes to System.Net.Security.Tests project file so as to copy the KDC setup script and related conf files to output directory

### CODE REVIEW

File: src/System.Net.Security/tests/FunctionalTests/System.Net.Security.Tests.csproj [hunk 1]
```diff
@@ -9,6 +9,7 @@
     <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
     <ProjectGuid>{A55A2B9A-830F-4330-A0E7-02A9FB30ABD2}</ProjectGuid>
     <OutputType>Library</OutputType>
+	<UnsupportedPlatforms>OSX</UnsupportedPlatforms>
```
Comment (stephentoub) [inline src/System.Net.Security/tests/FunctionalTests/System.Net.Security.Tests.csproj]: Nit: indentation

File: src/System.Net.Security/tests/FunctionalTests/System.Net.Security.Tests.csproj [hunk 2]
```diff
@@ -85,6 +86,21 @@
...
+	   <CopyToOutputDirectory>Always</CopyToOutputDirectory>
+	</None>
+	<None Include="..\Scripts\kdc.conf.ubuntu">
+	   <CopyToOutputDirectory>Always</CopyToOutputDirectory>
+	</None>
+	<None Include="..\Scripts\kdc.conf.centos">
+	   <CopyToOutputDirectory>Always</CopyToOutputDirectory>
+	</None>
+  </ItemGroup>
+
```
Comment (stephentoub) [inline src/System.Net.Security/tests/FunctionalTests/System.Net.Security.Tests.csproj]: Nit: indentation on all of this

File: src/System.Net.Security/tests/Scripts/setup-kdc.sh [hunk 3]
```diff
@@ -0,0 +1,121 @@
+#!/bin/bash
```
Comment (stephentoub) [inline src/System.Net.Security/tests/Scripts/setup-kdc.sh]: Is there any danger that something in this script will mess with a real configuration someone already has on their machine?  If so, before overwriting / deleting / changing / etc., we should prompt to make sure the caller is ok with the changes.

File: src/System.Net.Security/tests/Scripts/setup-kdc.sh [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+OS=`cat /etc/os-release | grep "^ID=" | sed 's/ID=//g' | sed 's/["]//g' | awk '{print $1}'`
+echo "Operating System: ${OS}"
+
+realm="TEST.COREFX.NET"
+
+principal1="HOST/host.test.corefx.net"
+principal2="HTTP"
+krb_user="krb_user"
+password="password"
```
Comment (stephentoub) [inline src/System.Net.Security/tests/Scripts/setup-kdc.sh]: If someone runs this script, is there any danger that it'll open their machine up to any kind of attack? e.g. could it give someone with these fake creds access to anything on their machine?

File: src/System.Net.Security/tests/Scripts/setup-kdc.sh [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+	# To overcome this and make file name expansion under sudoer/superuser, rm is run under a separate bash 
+	sudo bash -c "rm -f ${database_files}"
+
+	# Create/copy krb5.conf and kdc.conf
+	echo "Copying krb5.conf and kdc.conf.."
+	sudo /bin/cp ${krb_conf} ${krb_conf_location}
+	sudo /bin/cp ${kdc_conf} ${kdc_conf_location}
+	
+	echo "Creating KDC database for realm ${realm}.."
+	sudo ${kdb5_util} create -r ${realm} -P ${password} -s
```
Comment (stephentoub) [inline src/System.Net.Security/tests/Scripts/setup-kdc.sh]: What if one of these steps fails?  Should we be checking for process exit codes and erroring out if it's not successful?

File: src/System.Net.Security/tests/Scripts/setup-kdc.sh (hunk 3 above)
Comment (bartonjs) [inline src/System.Net.Security/tests/Scripts/setup-kdc.sh]: Yeah, the amount of machine-state altering that goes on here makes me nervous.

**Comment by bartonjs** [comment]
cc @morganbr

File: src/System.Net.Security/tests/Scripts/setup-kdc.sh [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+			then
+				echo "Error occurred during installation, aborting"
+				exit 1
+			fi
+		else
+			echo "krb5-kdc already installed.."
+			exit 0
+		fi
+		
+		configure_kdc
```
Comment (kapilash) [inline src/System.Net.Security/tests/Scripts/setup-kdc.sh]: If user already has krb5 installed this script will be a no-op. In such cases, the tests may fail (the user's krb5.conf may not contain the principals needed for the test)

Comment (stephentoub) [inline src/System.Net.Security/tests/Scripts/setup-kdc.sh]: Is there a way to prompt the user to overwrite their set up if that's actually what they want?

What about an uninstall, that undoes all of the configuration done in the script?

File: src/System.Net.Security/tests/Scripts/setup-kdc.sh (hunk 4 above)
Comment (morganbr) [inline src/System.Net.Security/tests/Scripts/setup-kdc.sh]: +1 on @stephentoub's question. I'm not convinced creating even low privilege accounts is a safe thing to do

File: src/System.Net.Security/tests/Scripts/setup-kdc.sh (hunk 3 above)
Comment (vijaykota) [inline src/System.Net.Security/tests/Scripts/setup-kdc.sh]: The script is meant to be part of automated runs (see #6139)

File: src/System.Net.Security/tests/Scripts/setup-kdc.sh (hunk 4 above)
Comment (vijaykota) [inline src/System.Net.Security/tests/Scripts/setup-kdc.sh]: This will enable getting a Kerberos ticket for the TEST.COREFX.NET realm but there won't be any server that would trust that ticket.

Comment (vijaykota) [inline src/System.Net.Security/tests/Scripts/setup-kdc.sh]: @morganbr can also chime in if we have missed something

File: src/System.Net.Security/tests/Scripts/setup-kdc.sh (hunk 6 above)
Comment (vijaykota) [inline src/System.Net.Security/tests/Scripts/setup-kdc.sh]: Prompting will not be possible due to the automated nature.

For uninstall, @stephentoub should we add an argument to the script? Then in #6139 should we also do an uninstall each time? (Note that always cleaning up will add to the time when devs run NegotiateStream tests locally)

File: src/System.Net.Security/tests/Scripts/setup-kdc.sh (hunk 3 above)
Comment (ellismg) [inline src/System.Net.Security/tests/Scripts/setup-kdc.sh]: Use `#!/usr/bin/env bash` instead.

File: src/System.Net.Security/tests/Scripts/setup-kdc.sh (hunk 5 above)
Comment (rahulkotecha) [inline src/System.Net.Security/tests/Scripts/setup-kdc.sh]: agreed. will handle failures in the script

File: src/System.Net.Security/tests/Scripts/setup-kdc.sh (hunk 6 above)
Comment (stephentoub) [inline src/System.Net.Security/tests/Scripts/setup-kdc.sh]: > Prompting will not be possible due to the automated nature.

That's not necessarily true.  There are many utilities that prompt, for example, unless you pass a /force or /Y or some such flag that says "I know what I'm doing, I automatically say 'yes' to any prompt, so don't prompt me".

> For uninstall, @stephentoub should we add an argument to the script?

Yes, please.

> Then in #6139 should we also do an uninstall each time?

If running the tests does an automatic install, then they should also automatically uninstall.  Otherwise, running the tests leaves the box in a state different than it started.  It should also be possible to run the uninstall manually and clean up any vestiges of the previous install.

> Note that always cleaning up will add to the time when devs run NegotiateStream tests locally

We should make it so that it only uninstalls if it needed to install.  That way, if a developer is working on this library and has manually run the install script, no extra install/uninstall time is taken in each run.  If it's someone who's just running a full test pass or otherwise running the tests locally, and they don't have things set up appropriately, the tests will do the install and uninstall and their box will (hopefully) be left clean.

File: src/System.Net.Security/tests/Scripts/setup-kdc.sh (hunk 4 above)
Comment (stephentoub) [inline src/System.Net.Security/tests/Scripts/setup-kdc.sh]: > but there won't be any server that would trust that ticket.

Including _this_ machine?  My concern isn't that some other machine would trust these credentials, but rather than this machine would trust them, and that, for example, after I've run this script on my Linux machine that someone else can use these wide-open credentials to gain some kind of access, even limited access, to my Linux machine.

Comment (vijaykota) [inline src/System.Net.Security/tests/Scripts/setup-kdc.sh]: Right.. I meant even this machine. @rahulkotecha there is no home directory etc for this user but you should verify if telnet, SSH etc are kerberized by default and can somehow allow a login

File: src/System.Net.Security/tests/Scripts/setup-kdc.sh (hunk 6 above)
Comment (vijaykota) [inline src/System.Net.Security/tests/Scripts/setup-kdc.sh]: Makes sense. Thanks for the idea about force switch

File: src/System.Net.Security/tests/Scripts/setup-kdc.sh (hunk 3 above)
Comment (bartonjs) [inline src/System.Net.Security/tests/Scripts/setup-kdc.sh]: @vijaykota I wouldn't expect this to succeed as a part of automated runs.  The first call to sudo is likely going to password prompt.  If the process's stdin makes it to the keyboard that'll hang.  If it's a static channel it'll fail.  Neither of those results in success.

I think that the tests will just need to use [ConditionalFact], not run on most people's machines, and we'll probably want to come up with a way of having a test fail if the condition is false and it's an official machine.

File: src/System.Net.Security/tests/Scripts/kdc.conf.ubuntu [hunk 7]
```diff
@@ -0,0 +1,16 @@
+[kdcdefaults]
+    kdc_ports = 750,88
+
+[realms]
+    TEST.COREFX.NET = {
+        database_name = /var/lib/krb5kdc/principal
+        admin_keytab = FILE:/etc/krb5kdc/kadm5.keytab
+        acl_file = /etc/krb5kdc/kadm5.acl
+        key_stash_file = /etc/krb5kdc/stash
+        kdc_ports = 750,88
```
Comment (bartonjs) [inline src/System.Net.Security/tests/Scripts/kdc.conf.ubuntu]: Is this line required here? It feels like it should be just using the equivalent one from [kdcdefaults], no?

File: src/System.Net.Security/tests/Scripts/kdc.conf.ubuntu [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+        max_life = 10h 0m 0s
+        max_renewable_life = 7d 0h 0m 0s
+        master_key_type = des3-hmac-sha1
+        supported_enctypes = aes256-cts:normal arcfour-hmac:normal des3-hmac-sha1:normal des-cbc-crc:normal des:normal des:v4 des:norealm des:onlyrealm des:afs3
+        default_principal_flags = +preauth
```
Comment (bartonjs) [inline src/System.Net.Security/tests/Scripts/kdc.conf.ubuntu]: A lot of lines that are present in the ubuntu file are missing from the centos one.  Why?

File: src/System.Net.Security/tests/Scripts/setup-kdc.sh (hunk 4 above)
Comment (rahulkotecha) [inline src/System.Net.Security/tests/Scripts/setup-kdc.sh]: Verified on Ubuntu, Debian, CentOS and RHEL. 
SSH, telnet are not kerberized by default, hence won't allow the kdc users to login through either of there protocols.

File: src/System.Net.Security/tests/Scripts/setup-kdc.sh (hunk 6 above)
Comment (rahulkotecha) [inline src/System.Net.Security/tests/Scripts/setup-kdc.sh]: @stephentoub we are not actually overwriting user's setup. If the kdc is already installed, we are assuming that the machine is already setup with required users for the testing (through earlier runs of the script). If this assumption is indeed correct, then we might not require the force switch. 
Let me know if this assumption is correct.

File: src/System.Net.Security/tests/Scripts/kdc.conf.ubuntu (hunk 7 above)
Comment (rahulkotecha) [inline src/System.Net.Security/tests/Scripts/kdc.conf.ubuntu]: This is default kdc.conf file generated with the KDC installation on Ubuntu/Debian.

File: src/System.Net.Security/tests/Scripts/kdc.conf.ubuntu (hunk 8 above)
Comment (rahulkotecha) [inline src/System.Net.Security/tests/Scripts/kdc.conf.ubuntu]: These are default kdc.conf files generated by the MIT KDC installation. Yes, the file on CentOS/RHEL is significantly different than that in Ubuntu/Debian out of the box.

File: src/System.Net.Security/tests/Scripts/setup-kdc.sh (hunk 6 above)
Comment (vijaykota) [inline src/System.Net.Security/tests/Scripts/setup-kdc.sh]: @rahulkotecha the force switch is also to let a dev running script outside CI to see some prompts before KDC config. Having the force switch allows the script to be run without prompts from test class setup

File: src/System.Net.Security/tests/Scripts/setup-kdc.sh (hunk 3 above)
Comment (vijaykota) [inline src/System.Net.Security/tests/Scripts/setup-kdc.sh]: @bartonjs this is taken care of in the Jenkins setup. You can see #6139 for some of the other points about conditionally running the tests.

File: src/System.Net.Security/tests/Scripts/setup-kdc.sh (hunk 4 above)
Comment (vijaykota) [inline src/System.Net.Security/tests/Scripts/setup-kdc.sh]: Thanks for confirming, @rahulkotecha They probably all need passwd entry that maps to the Kerberos principal. One minor change we should make is to use non-standard service principal names. eg. change HTTP to TestHTTP etc. This will further reduce probability of a regular service using the service principals we are creating.

cc: @joshfree 
@morganbr , @joshfree, can you recommend some Unix security expert who can validate our thinking?

File: src/System.Net.Security/tests/Scripts/setup-kdc.sh [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+
+OS=`cat /etc/os-release | grep "^ID=" | sed 's/ID=//g' | sed 's/["]//g' | awk '{print $1}'`
+echo "Operating System: ${OS}"
+
+realm="TEST.COREFX.NET"
+
+principal1="TESTHOST/testfqdn.test.corefx.net"
+principal2="TESTHTTP"
+krb_user="krb_user"
+password="password"
```
Comment (stephentoub) [inline src/System.Net.Security/tests/Scripts/setup-kdc.sh]: I know steps have been taken to mitigate concerns related to this, but are there additional ones we can take?  For example, do these credentials need to be hard coded, or could the setup script generate them randomly?

File: src/System.Net.Security/tests/Scripts/setup-kdc.sh [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+usage()
+{
+	echo "Usage: ${PROGNAME} [-y|--yes] [-u|--uninstall]";
+}
+
+# Cleanup config files and uninstall KDC
+clean_up()
+{
+	echo "Stopping KDC.."
+	if pgrep krb5kdc 2> /dev/null; then sudo killall krb5kdc ; fi
```
Comment (stephentoub) [inline src/System.Net.Security/tests/Scripts/setup-kdc.sh]: Rather than the script itself using ```sudo```, would it work to require the invoker of the script to use ```sudo``` to invoke it?  That would further prevent any concerns around the script going wild, as it wouldn't implicitly try to elevate.

File: src/System.Net.Security/tests/Scripts/setup-kdc.sh [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+		"centos" | "rhel")
+			kdc_conf_location="/var/kerberos/krb5kdc/kdc.conf"
+			yum list installed krb5-server >/dev/null 2>&1
+			if [ $? -eq 0 ]; then
+				echo "Uninstalling krb5-server"
+				sudo yum -y remove krb5-server
+			fi
+		;;
+
+		*)
```
Comment (stephentoub) [inline src/System.Net.Security/tests/Scripts/setup-kdc.sh]: Should we print something out if the OS wasn't recognized?

File: src/System.Net.Security/tests/Scripts/setup-kdc.sh [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+	# To overcome this and make file name expansion under sudoer/superuser, rm is run under a separate bash
+	sudo bash -c "rm -f ${database_files}"
+	
+	# Create/copy krb5.conf and kdc.conf
+	echo "Copying krb5.conf and kdc.conf.."
+	sudo /bin/cp ${krb_conf} ${krb_conf_location} || \
+	error_exit "Cannot copy ${krb_conf} to ${krb_conf_location}"
+	
+	sudo /bin/cp ${kdc_conf} ${kdc_conf_location} || \
+	error_exit "Cannot copy ${kdc_conf} to ${kdc_conf_location}"
```
Comment (stephentoub) [inline src/System.Net.Security/tests/Scripts/setup-kdc.sh]: What's the state of the machine if one of these errors out?  Will uninstall successfully undo any changes made thus far, even if the install wasn't complete?  I just want to make sure that a dev can get back to a clean state if something goes wrong part way through install.

File: src/System.Net.Security/tests/Scripts/setup-kdc.sh [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+}
+eval set -- "$TEMP"
+uninstall=0
+force=0
+while true; do
+	case $1 in
+        -y|--yes) force=1; shift ;;
+        -u|--uninstall) uninstall=1; shift;;
+        --) shift; break;;
+        *) echo "Usage"; exit 1;;
```
Comment (stephentoub) [inline src/System.Net.Security/tests/Scripts/setup-kdc.sh]: should this call ```usage()```?

**Comment by stephentoub** [comment]
Thanks, @rahulkotecha.  A few more comments/questions, but overall LGTM.

@mmitche, could you take a peek as well?

File: src/System.Net.Security/tests/Scripts/setup-kdc.sh (hunk 12 above)
Comment (rahulkotecha) [inline src/System.Net.Security/tests/Scripts/setup-kdc.sh]: Yes, there is cleanup procedure in place which, upon occurrence of an error, will revert the changes made to the system thus far (config files, installation of kdc, etc.)

File: src/System.Net.Security/tests/Scripts/setup-kdc.sh (hunk 13 above)
Comment (rahulkotecha) [inline src/System.Net.Security/tests/Scripts/setup-kdc.sh]: Yes, will fix.

File: src/System.Net.Security/tests/Scripts/setup-kdc.sh (hunk 11 above)
Comment (rahulkotecha) [inline src/System.Net.Security/tests/Scripts/setup-kdc.sh]: Yes, will add an echo statement.

File: src/System.Net.Security/tests/Scripts/setup-kdc.sh (hunk 9 above)
Comment (vijaykota) [inline src/System.Net.Security/tests/Scripts/setup-kdc.sh]: We couldn't think of a way to create a random password that the test case also would be aware of. @stephentoub, if the test case is always running elevated, we can store a random password in a well-known file which requires root access for reading. But if we expect the 3rd party dev experience running tests outside CI to run non-elevated, this would be an issue.

**Comment by rahulkotecha** [comment]
incorporated all comments received so far and squashed.

**Comment by vijaykota** [comment]
@mmitche please let us know if the changes look ok.

**Comment by vijaykota** [comment]
@rahulkotecha there are some conflicts (probably because setup-kdc.sh has been checked in via Priya's commit)

File: src/System.Net.Security/tests/Scripts/setup-kdc.sh (hunk 9 above)
Comment (stephentoub) [inline src/System.Net.Security/tests/Scripts/setup-kdc.sh]: > if the test case is always running elevated, we can store a random password in a well-known file which requires root access for reading. But if we expect the 3rd party dev experience running tests outside CI to run non-elevated, this would be an issue.

Right now the password is in a web page for the whole world to see. Even if a random one is written to a local file that doesn't require root access, isn't that still much better?

Comment (vijaykota) [inline src/System.Net.Security/tests/Scripts/setup-kdc.sh]: Agreed :) @rahulkotecha I think we should make this change

File: src/System.Net.Security/tests/Scripts/setup-kdc.sh (hunk 10 above)
Comment (vijaykota) [inline src/System.Net.Security/tests/Scripts/setup-kdc.sh]: cc: @kapilash I believe @kapilash tried this earlier but was having trouble when running tests locally using "corerun xunit.console.netcore.exe"

**Comment by mmitche** [comment]
Changes look good to me

@dotnet-bot test this please

**Comment by ellismg** [comment]
Is there a reason we can't just stand up a Kerberos in Azure and depend on that, instead of requiring folks that want to run these tests do all the work to set things up?

File: src/System.Net.Security/tests/Scripts/setup-kdc.sh (hunk 9 above)
Comment (vijaykota) [inline src/System.Net.Security/tests/Scripts/setup-kdc.sh]: Filed #6329 for this to take care of in a followup PR

**Comment by vijaykota** [comment]
> Is there a reason we can't just stand up a Kerberos in Azure 

@ellismg, that will probably be the eventual solution similar to say - HttpClient - since the Windows tests also need this for Kerberos testing. But we haven't been able to prototype this multi-machine solution so using this script to unblock ourselves till then.
